src/components/progressLinear/progress-linear.js

```diff
@@ -59,57 +79,112 @@ function MdProgressLinearDirective($$rAF, $mdConstant, $mdTheming) {
 
-    attr.$observe('value', function(value) {
-      if (attr.mdMode == 'query') {
-        return;
-      }
+    validateMode();
+    watchAttributes();
 
-      var clamped = clamp(value);
-      element.attr('aria-valuenow', clamped);
-      bar2Style[$mdConstant.CSS.TRANSFORM] = transforms[clamped];
-    });
+    /**
+     * Watch the value, md-buffer-value, and md-mode attributes
+     */
+    function watchAttributes() {
+      attr.$observe('value', function(value) {
+        var percentValue = clamp(value);
+        element.attr('aria-valuenow', percentValue);
 
-    attr.$observe('mdBufferValue', function(value) {
-      bar1Style[$mdConstant.CSS.TRANSFORM] = transforms[clamp(value)];
-    });
+        if (mode() != MODE_QUERY) animateIndicator(bar2, percentValue);
+      });
 
-    $$rAF(function() {
-      container.addClass('md-ready');
-    });
-  }
+      attr.$observe('mdBufferValue', function(value) {
+        animateIndicator(bar1, clamp(value));
+      });
 
-  function clamp(value) {
-    if (value > 100) {
-      return 100;
+      attr.$observe('mdMode',function(mode){
+        switch( mode ) {
+          case MODE_QUERY:
+          case MODE_BUFFER:
+          case MODE_DETERMINATE:
+          case MODE_INDETERMINATE:
+            container.removeClass('ng-hide');
+
+            // Inject class selector instead of attribute selector
+            // (@see layout.js changes for IE performance issues)
+
+            if ( lastMode ) container.removeClass( lastMode );
+                 lastMode = "md-mode-" + mode;
+            if ( lastMode ) container.addClass( lastMode );
+
+            break;
+          default:
+            container.addClass('ng-hide');
+            if ( lastMode ) {
+              container.removeClass( lastMode );
+              lastMode = undefined;
+            }
+            break;
+        }
+      });
     }
 
-    if (value < 0) {
-      return 0;
+    /**
+     * Auto-defaults the mode to either `determinate` or `indeterminate` mode; if not specified
+     */
+    function validateMode() {
+      if ( angular.isUndefined(attr.mdMode) ) {
+        var hasValue = angular.isDefined(attr.value);
+        var mode = hasValue ? MODE_DETERMINATE : MODE_INDETERMINATE;
+        var info = "Auto-adding the missing md-mode='{0}' to the ProgressLinear element";
+
+        $log.debug( $mdUtil.supplant(info, [mode]) );
+
+        element.attr("md-mode",mode);
+        attr['mdMode'] = mode;
+      }
     }
 
-    return Math.ceil(value || 0);
-  }
-}
+    /**
+     * Is the md-mode a valid option?
+     */
+    function mode() {
+      var value = attr.mdMode;
+      if ( value ) {
+        switch(value) {
+          case MODE_DETERMINATE:
+          case MODE_INDETERMINATE:
+          case MODE_BUFFER:
+          case MODE_QUERY:
+            break;
+          default:
+            value = undefined;
+            break;
+        }
+      }
```

Comment:
Just curious, is this wrapping `if( value )` necessary? Shouldn't the `default` case catch that and set it to `undefined`?

Or is this just a way to optimize the case of it being undefined so you don't have to run the switch?